app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,332 @@
+        locationManager.removeUpdates(locationUpdateListener)
+        locationUpdates.postValue(null)
+        isUpdating = false
+        return true
+    }
+
+    fun isLocationEnabled(): Boolean {
+        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
+    }
+
+    fun hasSufficientAccuracy(): Boolean {
+        return currentLocation?.let {
+            it.hasAccuracy() && it.accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING
+        } ?: false
+    }
+
+    private fun hasLocationPermissions(): Boolean {
+
+        val fineLocationPermission = ContextCompat.checkSelfPermission(
+            context,
+            android.Manifest.permission.ACCESS_FINE_LOCATION
+        )
+        val coarseLocationPermission = ContextCompat.checkSelfPermission(
+            context,
+            android.Manifest.permission.ACCESS_COARSE_LOCATION
+        )
+        return (fineLocationPermission == PackageManager.PERMISSION_GRANTED ||
+                coarseLocationPermission == PackageManager.PERMISSION_GRANTED)
+    }
+}
+
+enum class Accuracy {
+    GOOD,
+    BAD,
+    NONE
+}
+
+fun Location?.accuracyStatus(): Accuracy {
+    if (this == null || !hasAccuracy()) {
+        return Accuracy.NONE
+    }
+    return if (accuracy < ValueHelper.MIN_ACCURACY_DEFAULT_SETTING) {
+        Accuracy.GOOD
+    } else {
+        Accuracy.BAD
+    }
+}
+
+class LocationDataCapturer(
+    private val userManager: UserManager,
+    private val locationUpdateManager: LocationUpdateManager,
+    private val treeTrackerDAO: TreeTrackerDAO
+) {
+    var convergenceWithinRange: Boolean = false
+
+    private val gson = GsonBuilder().serializeNulls().create()
+    private var locationsDeque: Deque<Location> = LinkedList<Location>()
+    var generatedTreeUuid: UUID? = null
+        private set
+    var convergence: Convergence? = null
+
+    private val locationObserver: Observer<Location?> = Observer {
+        it?.apply {
+
+            MainScope().launch(Dispatchers.IO) {
+                val locationData =
+                    LocationData(
+                        userManager.planterCheckinId,
+                        latitude,
+                        longitude,
+                        accuracy,
+                        generatedTreeUuid?.toString() ?: null,
+                        System.currentTimeMillis()
+                    )
+                val jsonValue = gson.toJson(locationData)
+                Timber.d("Inserting new location data $jsonValue")
+                treeTrackerDAO.insertLocationData(LocationDataEntity(jsonValue))
+            }
+
+            if (isInTreeCaptureMode() && !convergenceWithinRange) {
+
+                var evictedLocation: Location? = if (locationsDeque.size >= CONVERGENCE_DATA_SIZE)
+                    locationsDeque.pollFirst() else null
+                locationsDeque.add(it)
+
+                if (locationsDeque.size >= CONVERGENCE_DATA_SIZE) {
```

Comment:
This condition is already checked at line 175. What if this instead checks for evitedLocation nullablilty so it's safe to use it later on. That should remove it's !! on line 184

Reply:
It would not satisfy the primary intent that computation only occurs if the minimum size is reached. The very first computation would be missed if I base it solely on the nullability of `evictedLocation`.

Reply:
The nullability depends on the size so they should be the same. Shouldn't it?

Reply:
Actually the very first time the size reaches (say 5) the required minimum, the `evictedLocation` is still null.  The line 177 will move the size of `locationsDeque` to 5 from 4 and this will be the very first computation and subsequent ones will evict the head and add the to the tail.

I have removed the `!!` and now using `?` within the if condition.